microprofile/WFLY-10712_microprofile_metrics.asciidoc

```diff
@@ -0,0 +1,407 @@
+* Standalone Full HA Profile (186 additional metrics):
+** `undertow` - 6 metrics (for 1 additional `ajp-listener`)
+** `jgroups` - 180 metrics (for 1 ee channel)
+
+====== Name of WildFly vendor metrics
+
+The name of the MicroProfile metric is derived from WildFly's attribute name and the resource that defines it.
+
+The name of the metric is composed of the address of the resource using the "path style" (where `=` are replaced by `/`) and the attribute name separated by a `/`:
+
+----
+<resource address in "path" style>/<attribute name>
+----
+
+Note that the metric name has no leading `/`.
+
+So for example, the metric `bytes-sent` on the resource `/subsystem=undertow/server=default-server/http-listener=default` will be named `subsystem/undertow/server/default-server/http-listener/default/bytes-sent`
+
+This will be the name of the metric registered in MicroProfile VENDOR metric registry and can be used from the HTTP endpoints.
+
+However, this name will then be converted when exported to the Prometheus format according to the translation rules (section 3.2.1 of the MicroProfile Metrics 1.1.1 specification):
+
+----
+$ curl -v http://127.0.0.1:9990/metrics/vendor/subsystem/undertow/server/default-server/http-listener/default/bytes-sent
+< HTTP/1.1 200 OK
+...
+# HELP vendor:subsystem_undertow_server_default_server_http_listener_default_bytes_sent_bytes The number of bytes that have been sent out on this listener
+# TYPE vendor:subsystem_undertow_server_default_server_http_listener_default_bytes_sent_bytes gauge
+vendor:subsystem_undertow_server_default_server_http_listener_default_bytes_sent_bytes 0.0
+...
+----
+
+===== Implementation Issues
+
+====== Invalid WildFly Metrics Registration
+
+Some resources in WildFly and WildFly Core codebase improperly registered runtime read-only attributes as metrics. This is tracked by https://issues.jboss.org/browse/WFCORE-4173[WFCORE-4173] and https://issues.jboss.org/browse/WFLY-11212[WFLY-11212]
+and will be resolved before this feature is integrated.
+
+====== Complex WildFly Metrics Are Not Supported
+
+Resources may return arbitrarily complex return types for metrics. Most complex example I found in the codebase is http://wildscribe.github.io/WildFly/14.0/core-service/platform-mbean/type/memory/index.html#attr-heap-memory-usage[ core-service=platform-mbean
+type=memory#heap-memory-usage].
+The registration code in `micrprofile-smallrye-metrics` will only register MicroProfile Metrics for simple numerical `ModelType`.
+
+====== WildFly Metrics Are Registered As Gauge
+
+As there is no way to determine whether a WildFly metric corresponds to a https://github.com/eclipse/microprofile-metrics/blob/1.1.1/api/src/main/java/org/eclipse/microprofile/metrics/Gauge.java[Gauge] or a https://github.com/eclipse/microprofile-metrics/blob/1.1.1/api/src/main/java/org/eclipse/microprofile/metrics/Counter.java[Counter], they
+will all be registered in the MicroProfile vendor registry as *Gauge*.
```

Comment:
Perhaps as a future enhancement an arbitrary attribute descriptor can be used to allow metrics to be classified, defaulting to 'gauge'.

I'm referring to "Arbitrary Descriptors" in http://docs.wildfly.org/14/Admin_Guide.html#Description_of_the_Management_Model